### ISSUE TITLE
Vulkan format of AMF_SURFACE_NV12

### ISSUE THREAD

**Issue by maxest** [GPUOpen-LibrariesAndSDKs/AMF#232]
I decoded an image and on the decoded data I call:
`amf::AMFSurfacePtr surface( amfOutputData );`
This surface is a Vulkan surface (I made sure amfOutputData is of Vulkan memory type).
The format of that surface is NV12.
I've tried copying this surface to my own Vulkan image (I create AMF surface off my Vulkan image using `CreateSurfaceFromVulkanNative`). However, I always get "src and dst should have same format" when I call `CopySurfaceRegion`.
The Vulkan format of my image that I think is closest to NV12 is VK_FORMAT_G8_B8R8_2PLANE_420_UNORM.

To sum up: How can I copy NV12 AMF surface to a Vulkan image (or buffer) of mine and what format do I need?

**Comment by MikhailAMD** [comment]
At this point AMF allocates NV12 with a private Vulkan format. So if you want a copy, use surface->Duplicate(). 
You can use individual AMF planes and internal Vulkan image views to use this surface in your rendering:
VkImageView hY=((AMFVulkanView*)surface->GetPlaneAt(0)->GetNative())->hView; // VK_FORMAT_R8_UINT
VkImageView hUV=((AMFVulkanView*)surface->GetPlaneAt(1)->GetNative())->hView; // VK_FORMAT_R8G8_UINT

you can convert plane type (create a different Vulkan view like UNORM) by AMFCompute::ConvertPlaneToPlane()

or you can manually create your own views (like UNORM) by using AMFVulkanView::pSurface->hImage

**Comment by maxest** [comment]
So, I can get VkImageView of the planes, but not VkImage?

To be more precise. Given that I have an AMF Vulkan image in NV12 format, and I have my custom two images, one R8_UNORM and the other R8G8_UNORM, I'd like to "copy/convert" the former into the latters.

**Comment by MikhailAMD** [comment]
In the last line of the previous comment I did show how to get VkImage:
VkImage hImage=((AMFVulkanView*)surface->GetPlaneAt(0)->GetNative())->pSurface->hImage;

But I don't think you can use this image in vkCmdCopyImage to copy individual planes because you would need to fill in aspectMask and you don't have private values for this field. 

Two options: 

You can wrap each your image into AMFSurface individually (AMFContext1::CreateSurfaceFromVulkanNative()) and use AMFCompute::CopyPlane()

Or use AMF-created views directly in your shaders. 

In any case don't forget about synchronization: check AMFVulkanSurface::Sync structure for hSemaphore and it state bSubmitted. See comments in VulkanAMF.h

**Comment by maxest** [comment]
Ah, I missed that. Thanks.

Isn't it enough to pass VK_IMAGE_ASPECT_PLANE_0_BIT as aspect mask? (after https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdCopyImage.html)

You mentioned Duplicate.
Can I call Duplicate on an AMF NV12 image, and as ppData to Duplicate pass my image, wrapped in AMFVulkanSurface, in format R8? Would the data be laid out like: linear luma first, then linear packed UV?

**Comment by MikhailAMD** [comment]
I doubt VK_IMAGE_ASPECT_PLANE_0_BIT would work, again this is private extension format and private aspects enum values.

Duplicate() copies to internal allocation only. 

You can use the last variant: wrap your images and call CopyPlane() for each.

**Comment by maxest** [comment]
I'm trying your suggestion. Unfortunately, seems like the driver is crashing when it executes CopyPlane. The system hangs and screen goes black for a few secs. Surprisingly, the function returns with AMF_OK...

This is the code:
`
	amf::AMFVulkanSurface amfSurf1 = {};
	amfSurf1.cbSizeof = sizeof(amf::AMFVulkanSurface);
	amfSurf1.hImage = vkPlaneY.Handle();
	amfSurf1.eUsage = amf::AMF_SURFACE_USAGE_DEFAULT;
	amfSurf1.hMemory = vkPlaneY.GetMemory();
	amfSurf1.iSize = vkPlaneY.m_memorySize;
	amfSurf1.eFormat = VK_FORMAT_R8_UNORM;
	amfSurf1.iWidth = 1920;
	amfSurf1.iHeight = 1080;
	amfSurf1.Sync.cbSizeof = sizeof(amfSurf1.Sync);
	amfSurf1.Sync.hSemaphore = VK_NULL_HANDLE;
	amfSurf1.Sync.bSubmitted = false;

	amf::AMFVulkanSurface amfSurf2 = {};
	amfSurf2.cbSizeof = sizeof(amf::AMFVulkanSurface);
	amfSurf2.hImage = vkPlaneUV.Handle();
	amfSurf2.eUsage = amf::AMF_SURFACE_USAGE_DEFAULT;
	amfSurf2.hMemory = vkPlaneUV.GetMemory();
	amfSurf2.iSize = vkPlaneUV.m_memorySize;
	amfSurf2.eFormat = VK_FORMAT_R8G8_UNORM;
	amfSurf2.iWidth = 1920 / 2;
	amfSurf2.iHeight = 1080 / 2;
	amfSurf2.Sync.cbSizeof = sizeof(amfSurf2.Sync);
	amfSurf2.Sync.hSemaphore = VK_NULL_HANDLE;
	amfSurf2.Sync.bSubmitted = false;

	{
		amf::AMFSurfacePtr surface( amfOutputData );
		amf::AMFPlane * planeY = surface->GetPlane( amf::AMF_PLANE_Y );
		amf::AMFPlane * planeUV = surface->GetPlane( amf::AMF_PLANE_UV );

		amf::AMFSurfacePtr surfaceY;
		res = context1->CreateSurfaceFromVulkanNative( &amfSurf1, &surfaceY, NULL );
		amf::AMFSurfacePtr surfaceUV;
		res = context1->CreateSurfaceFromVulkanNative( &amfSurf2, &surfaceUV, NULL );

		amf_size originSrc[3] = { 0, 0, 0};
		amf_size originDst[3] = { 0, 0, 0};
		amf_size regionY[3] = { 1920, 1080, 1};
		amf_size regionUV[3] = { 1920 / 2, 1080 / 2, 1};

		amf::AMFComputePtr pCompute;
		res = context->GetCompute(amf::AMF_MEMORY_VULKAN, &pCompute);
		res = pCompute->CopyPlane( planeY, originSrc, regionY, surfaceY->GetPlane(amf::AMF_PLANE_PACKED), originDst );
		res = pCompute->CopyPlane( planeUV, originSrc, regionUV, surfaceUV->GetPlane(amf::AMF_PLANE_PACKED), originDst );
	}
`

**Comment by MikhailAMD** [comment]
- What are return codes for each CreateSurfaceFromVulkanNative() call?
- What is in debug output? 
Not related: keep in mind that &amfSurf1 and &amfSurf2 are treated as handles so you have to keep the pointers alive till wrap surfaces  are used/released.

**Comment by maxest** [comment]
So it was an error on my side. The images were not really created (image handle was, but no memory was bound.
Things seem to be working.

I have one last question regarding sync. I noticed that my perf timing is skewed cause it was showing decode is taking ~0.5 ms. So I added FinishQueue and FlushQueue after the CopyPlane(s) and now it shows ~2 ms what is closer to what I'd expect. This should take care of syncing all the decoding process, right?

But if I understand correctly there is another problem. I'm doing rendering of the decoded image on the main thread and all the decoding is carried out on a separate thread. This creates a problem that the image that I am copying to, from AMF, can also be used at the same time by the main thread. Could this be a problem?

**Comment by MikhailAMD** [comment]
Yes, if you want CPU sync, FlushQueue/FinishQueue is the correct sequence.
  
The decoded surface /image will be reused by decoder only after you release the surface pointer. You can do it in any thread. And you don't have to keep it after FinishQueue(). 
The surface/image that you allocated and wrapped in AMF will not be used/ touched after FinishQueue(), you are free to do with it whatever you want.

Note, that while FinishQueue gives you proper CPU sync the most optimal way is to use provided semaphores and sync on GPU only. AMF decode and copy operations will synchronize themselves with these semaphores even if decode surface is released. It is up to you to use it or not.

**Comment by maxest** [comment]
I have a question regarding performance. This is my decode in pseudocode:
```
CreateBufferFromHostNative

SubmitInput
QueryOutput

CopyPlaneY
CopyPlaneUV
Compute:FinishQueue
```
I have measured the time of it and it takes ~2.3 ms on my 5700 XT.
I noticed though there is this AMF_VIDEO_DECODER_MODE_LOW_LATENCY mode. So I activate it on the decoder. Now the CPU time of all those operations is ~6.5 ms. Where is this difference coming from?

We have "input-to-picture" latency calculator and seems like using AMF_VIDEO_DECODER_MODE_LOW_LATENCY indeed does reduce latency a little bit. From visual inspection it's hard to tell.

I thought the ~2.3 ms was accurate cause we also have NVIDIA decoder implemented in the same way (Submit, Query, copy result to our VK images) and it reports a very similar timing on a notebook's RTX 2080. Given the price diff between RTX 2080 and 5700 XT and the fact I am comparing a desktop 5700 XT with a notebook's RTX, I'd expect the timings to actually be between similar and 30-50% difference. But not almost 3x times.

**Comment by MikhailAMD** [comment]
AMF_VIDEO_DECODER_MODE_LOW_LATENCY only changes surface the reorder logic. If you know that your stream doesn't have B-frames or funny POCs with reordering, this flag tells decoder to output right after a frame is submitted. Please note that most of the time GPU has power management and there is no guarantee that decoder will provide smallest latency. Counterintuitively,  more the load, faster it works.
